src/QuLDE.jl

```diff
@@ -41,17 +58,86 @@ struct QuLDEParam{W,L,Q, VType <: GeneralMatrixBlock{1}, SType <: GeneralMatrixB
         VS2 = matblock(VS2)
         WS1 = matblock(WS1)
         WS2 = matblock(WS2)
-        n = Int(log2(k+1))
+        n = log2i(k+1)
         new{ComplexF64,L,n, typeof(V), typeof(VS1)}(k,t,C_tilda,D_tilda,N,V,VS1,VS2,WS1,WS2)
     end
 end
 
-function quldecircuit(blk::QuLDEParam, M::Matrix)
-    T = Int(log2(blk.k+1))
-    siz, = size(M)
-    nbit = Int(log2(siz))
+struct QuLDEnonUnitParam{L, W, Q, NL, Nbit, VType <: GeneralMatrixBlock{1}, SType <: GeneralMatrixBlock{Q}, LType <: GeneralMatrixBlock{NL}}
+    k::Int # Taylor expansion upto k
+    t::L # time
+    l::Int # input size of LType block set to 2
+    C_tilda::W #G1 as per the article
+    D_tilda::W #G2 as per the article
+    N::W
+    V::VType
+    VS1::SType # Q is input size of GMblock
+    VS2::SType # Q
+    WS1::SType # Q
+    WS2::SType # Q
+    VT::LType # NL is input size of GMblock
+    WT::LType # NL
+    F::Array{GeneralMatrixBlock{Nbit,Nbit,Complex{Float64},Array{Complex{Float64},2}},1} # Nbit is input size
+
+    function QuLDEnonUnitParam(k::Int,t::L,l::Int,prob::QuLDEProblem) where {L}
+        C(m) = norm(prob.u0)*((opnorm(prob.A)*t*2)^(m))/factorial(m) # alphas equal 1/2
+        D(n) = norm(prob.b)*(opnorm(prob.A)*t*2)^(n-1)*t/factorial(n)
+        nbit, = size(prob.u0)
+        nbit = log2i(nbit)
+        Mu = prob.A/opnorm(prob.A)
+        B1 = 1/2*(Mu + Mu')
+        B2 = -im/2*(Mu - Mu')
+        iden = Matrix{ComplexF64}(I,size(B1))
+        F = Array{GeneralMatrixBlock{nbit,nbit,Complex{Float64},Array{Complex{Float64},2}},1}(undef,4)
+        F[1] = matblock(B1 + im*sqrt(iden - B1*B1))
+        F[2] = matblock(B1 - im*sqrt(iden - B1*B1))
+        F[3] = matblock(im*B2 - sqrt(iden - B2*B2))
+        F[4] = matblock(im*B2 + sqrt(iden - B2*B2))
+        #tested
+
+        C_tilda = 0
+        D_tilda = 0
+        for i in 1:k
+            C_tilda = C_tilda + C(i)
+            D_tilda = D_tilda + D(i)
+        end
+        C_tilda = C_tilda + C(0)
+        N = sqrt(C_tilda+ D_tilda)
+        C_tilda = sqrt(C_tilda)
+        D_tilda = sqrt(D_tilda)
+        V = ComplexF64[C_tilda D_tilda; D_tilda -1*C_tilda]/N
+        V = matblock(V)
+        #tested
+        VS1 = rand(ComplexF64,2^k,2^k)
```

Comment:
so this is an example that it is not generic enough to let it run on CUDA. But let's do it later, make things work first. And you could also keep this in mind the following coding.

To make it generic enough for CUDA, we need at least to support a general floating point precision (which means the program can also use `Float32`)

Reply:
Is there an example I take a look at, to sort of better understand how to go about it?

Reply:
For example, you could just let all `ComplexF64` be a type parameter (if you read the implementation of YaoBlocks, we never use a specific type, so the code would be very generic).